sklearn/metrics/pairwise.py

```diff
@@ -870,6 +870,53 @@ def haversine_distances(X, Y=None):
     return DistanceMetric.get_metric("haversine").pairwise(X, Y)
 
 
+def paired_haversine_distances(X, Y):
+    """Compute the Haversine distance between paired samples in X and Y.
+
+    The Haversine (or great circle) distance is the angular distance between
+    two points on the surface of a sphere. The first coordinate of each point
+    is assumed to be the latitude, the second is the longitude, given
+    in radians. The dimension of the data must be 2.
+
+    .. math::
+       D(x, y) = 2\\arcsin[\\sqrt{\\sin^2((x1 - y1) / 2)
+                                + \\cos(x1)\\cos(y1)\\sin^2((x2 - y2) / 2)}]
+
+    Parameters
+    ----------
+    X : array-like of shape (n_samples_X, 2)
+        A numpy array of coordinates (latitude and longitude, in radians).
+
+    Y : array-like of shape (n_samples_Y, 2)
+        A numpy array of coordinates (latitude and longitude, in radians).
+
+    Returns
+    -------
+    distance : ndarray of shape (n_samples_X, )
+        The distances between the points in X and Y.
+
+    Notes
+    -----
+    As the Earth is nearly spherical, the haversine formula provides a good
+    approximation of the distance between two points of the Earth surface, with
+    a less than 1% error on average.
+    """
+    X, Y = check_paired_arrays(X, Y)
+    if not (len(X.shape) == len(Y.shape) == 2) or not (X.shape[1] == Y.shape[1] == 2):
+        raise ValueError(
+            "For paired distances and the metric `haversine`, "
+            "X and Y should both be of shape (n_samples, 2), "
+            "but X.shape is {0} and Y.shape is {1}".format(X.shape, Y.shape)
+        )
```

Comment:
Do you mind adding a test verifying that the ValueError is raised?

Reply:
Sorry, I see now that you added it already... I'd rather have an independent test checking the error message too... but I guess this is not essential...

Reply:
@cmarmo Does that mean you're ok with the test as is?

Reply:
I find more readable to have a separate test. But I think it is better to have another opinion, just to avoid double back and forth.
I'm labeling this one as "Waiting for review": let's hope someone will jump in quickly... :)